### PR TITLE
[Entity Layering] Added schema changes for entity layering

### PR DESCRIPTION
**Pull request by Ceiu**
CANDLEPIN-668: Added schema changes for entity layering

- Added the new tables for the layered entity work
- Removed the cp2_ prefix from new product and content tables
- Removed the "locked" and "entity_version" fields from the product
  and content entities
- Product and content entities are no longer immutable
- Removed the now-obsolete LongHashCodeBuilder
- Added the "namespace" field to product and content entities
- Updated ProductManager and ContentManager to no longer perform
  versioning logic, and instead support namespacing and layering
  logic
- Moved some curator logic from the OwnerContentCurator and
  OwnerProductCurators to more fitting curators
- Renamed the "locked" field on the Pool entity to "managed"
- Fixed a bug in PoolOpProcessor.setPoolQuantity which would
  skip applying the db lock on managed pool when updating the
  pool quantity
- Fixed a bug that would cause derived pools of managed pools
  to not inherit the managed state
- Removed the restriction on product and content deletion that
  would disallow the delete operation on any entity which is
  referenced by one or more parent products. The entity will
  now be automatically unlinked as part of the delete op.
- Updated refresh behavior to only fetch, map, and update
  entities which are related to or affected by the entities
  fetched from the upstream data source
- Refresh and import no longer remove "unused" product or
  content definitions

CANDLEPIN-671: Updated API to support entity layering

- Updated the API endpoint operation IDs to better indicate their
  behavior, and standardize their naming
- Updated product and content CRUD operations to support namespaces
- Removed the endpoints for listing of owners for a given product
  or content, as they no longer make sense in the context of
  namespaced entities

### CODE REVIEW

**Comment by Ceiu** [comment]
I'm torn on how to deal with this PR. Ripping up the schema requires logic changes; but unfortunately these logic changes are pretty core to how products and content work today. Dealing with it in this PR means also addressing a couple other tasks at the same time (updating the refresher logic and service layer) which are technically outside the original scope of this task.

The correct thing to do here is probably leave this PR up as a draft for review purposes, but not merge it until the rest of the code is present such that tests pass as expected, without the warning in place and the risk someone will forget to squash before the feature is merged.

**Comment by wottop** [comment]
Are we into feature branch territory?

**Comment by Ceiu** [comment]
This PR is blocked by the following:
- #4370 
- #4364

File: src/main/java/org/candlepin/controller/PoolConverter.java
Comment (nikosmoum) [inline src/main/java/org/candlepin/controller/PoolConverter.java:96]: Looking at the usage of this method, it is used in 2 places:
1) The Refresher
2) PoolRules, and looking at that call stack, it goes all the way to `public List<Pool> createAndEnrichPools(SubscriptionInfo sub)`, which is only used by unit tests, so it should be removed (along with all the intermediate methods).

Comment (nikosmoum) [reply src/main/java/org/candlepin/controller/PoolConverter.java:96]: Also, it may be a good idea to group pool management classes by package, to indicate what/where they're supposed to belong. For example PoolConverter looks like it should only ever be used by Refresh/Import, so maybe it should go somewhere under `org.candlepin.controller.refresher` instead of just `org.candlepin.controller`

**Review by nikosmoum — COMMENTED** [review]
Haven't gotten through the whole thing yet, but here's some comments for now

File: api/candlepin-api-spec.yaml
Comment (Ceiu) [reply api/candlepin-api-spec.yaml:7020]: Nope, just making an educated/hopeful guess and will be relying on the test period with Sat to catch any issues.

Though we're going to be in a bit of trouble if it is: It *can't* work with the new implementation.

File: src/main/java/org/candlepin/controller/PoolConverter.java
Comment (Ceiu) [reply src/main/java/org/candlepin/controller/PoolConverter.java:96]: I'll remove the `PoolRules` path, but I don't really want to think about reorganizing the classes I'm not working in at the moment. Should create a backlog task for that.

File: src/main/java/org/candlepin/controller/OwnerManager.java
Comment (Ceiu) [reply src/main/java/org/candlepin/controller/OwnerManager.java:184]: While you're right, this is a seldom hit endpoint and the delta will be small. A custom UUID-only query will skip object hydration and a possible follow-up query for attributes or something.

Keep an eye on this, though. If we're doing this in other places then we should consider a specific query for it. But given the mountain of testing I still have ahead of me, I'm not too keen on adding another pair of methods for a one-off.

File: src/main/java/org/candlepin/controller/PoolConverter.java
Comment (Ceiu) [reply src/main/java/org/candlepin/controller/PoolConverter.java:96]: After spending a couple hours poking at the test suite with the pool rules conversion removed, I'm opting to punt both to a separate task. The pool manager testing is an absolute mess. There are pool rules tests hanging out in `PoolManagerTest`, pool manager tests in `PoolCuratorTest`, and a bunch of unnecessary refresh loops for some reason. All of that would have been fine if PoolManagerTest wasn't an awful pile of clunky mocks which I really don't want to deal with in this already-XL sized PR.

As far as the line in question goes, I've updated it to use the namespace of the org from the subscription since it won't proceed without one anyway.

**Review by Januson — COMMENTED** [review]
Changes look good. I have just a few nitpicks.

File: src/test/java/org/candlepin/policy/PoolRulesTest.java
Comment (Ceiu) [reply src/test/java/org/candlepin/policy/PoolRulesTest.java:1034]: Removed

File: src/main/java/org/candlepin/hibernate/NullAsEmptyStringType.java
Comment (Ceiu) [reply src/main/java/org/candlepin/hibernate/NullAsEmptyStringType.java:118]: Fixed

**Review by joshmalbrecht — CHANGES_REQUESTED** [review]
Looks good! I just noticed some small things and had a couple questions

File: src/main/java/org/candlepin/async/tasks/OrphanCleanupJob.java
Comment (Ceiu) [reply src/main/java/org/candlepin/async/tasks/OrphanCleanupJob.java:63]: There's already a follow-up task to fully remove the orphan cleanup stuff: CP-676. This PR just removed the minimal amount for tests to pass.

File: src/main/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitor.java
Comment (Ceiu) [reply src/main/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitor.java:138]: Which two?

File: src/main/java/org/candlepin/resource/OwnerContentResource.java
Comment (Ceiu) [reply src/main/java/org/candlepin/resource/OwnerContentResource.java:162]: Yes; fixed

File: src/main/java/org/candlepin/resource/OwnerProductResource.java
Comment (Ceiu) [reply src/main/java/org/candlepin/resource/OwnerProductResource.java:158]: Fixed

File: src/main/java/org/candlepin/resource/OwnerProductResource.java
Comment (Ceiu) [reply src/main/java/org/candlepin/resource/OwnerProductResource.java:153]: Fixed

File: src/main/java/org/candlepin/resource/OwnerProductResource.java
Comment (Ceiu) [reply src/main/java/org/candlepin/resource/OwnerProductResource.java:254]: Fixed

File: src/main/java/org/candlepin/resource/OwnerProductResource.java
Comment (Ceiu) [reply src/main/java/org/candlepin/resource/OwnerProductResource.java:337]: Fixed

File: src/test/java/org/candlepin/controller/ProductManagerTest.java
Comment (Ceiu) [reply src/test/java/org/candlepin/controller/ProductManagerTest.java:567]: Deleted. This test has been obsoleted by no longer needing to re-roll product/content instances whenever they change.

File: src/main/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitor.java
Comment (joshmalbrecht) [reply src/main/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitor.java:138]: Line 137 and 138, although I can see this being a stylistic choice to combine the two lines

File: src/main/java/org/candlepin/async/tasks/OrphanCleanupJob.java
Comment (joshmalbrecht) [reply src/main/java/org/candlepin/async/tasks/OrphanCleanupJob.java:63]: Perfect!

**Review by joshmalbrecht — APPROVED** [review]

File: src/main/java/org/candlepin/controller/PoolManager.java
Comment (nikosmoum) [inline src/main/java/org/candlepin/controller/PoolManager.java:269]: Will this be done as part of another task?

**Review by nikosmoum — CHANGES_REQUESTED** [review]

File: src/main/java/org/candlepin/resource/ProductResource.java
Comment (Ceiu) [reply src/main/java/org/candlepin/resource/ProductResource.java:104]: - Consistency and avoiding future pain (endpoint lists all products from all namespaces)
- Max should be total number of all known RH products + sum of all custom products across all orgs (~10k total)

File: src/main/java/org/candlepin/controller/ContentManager.java
Comment (Ceiu) [reply src/main/java/org/candlepin/controller/ContentManager.java:134]: In this case, it's a statement of permission or capability, as in "it is not able to update the UUID, ID, or namespace".

I can update that to read "... method is not able to update ..."

File: src/main/java/org/candlepin/controller/PoolManager.java
Comment (Ceiu) [reply src/main/java/org/candlepin/controller/PoolManager.java:269]: This will be vetted and completed (if necessary) as part of CP-671

File: src/main/java/org/candlepin/resteasy/filter/StoreFactory.java
Comment (Ceiu) [reply src/main/java/org/candlepin/resteasy/filter/StoreFactory.java:310]: I can try, but if there are any failures or oddities, I'll defer its removal until we have time to do a proper investigation.

File: src/main/resources/db/changelog/20231003113535-entity_layering_schema.xml
Comment (Ceiu) [reply src/main/resources/db/changelog/20231003113535-entity_layering_schema.xml:81]: Should be. Everything is generally either 32 for IDs or 255 for arbitrary strings, and a quick glance at the comparison in a deployed Candlepin checked out.

```diff
-                              Table "public.cp2_products"
-         Column        |            Type             | Collation | Nullable | Default 
- ----------------------+-----------------------------+-----------+----------+---------
-  uuid                 | character varying(32)       |           | not null | 
-  created              | timestamp without time zone |           |          | 
-  updated              | timestamp without time zone |           |          | 
-  multiplier           | bigint                      |           |          | 
-  product_id           | character varying(32)       |           | not null | 
-  name                 | character varying(255)      |           | not null | 
-  entity_version       | bigint                      |           |          | 
-  locked               | integer                     |           |          | 
-  derived_product_uuid | character varying(32)       |           |          | 
- Indexes:
-     "cp2_products_pk" PRIMARY KEY, btree (uuid)
-     "cp2_product_entity_version" UNIQUE CONSTRAINT, btree (product_id, entity_version) DEFERRABLE
-     "cp2_products_idx1" btree (entity_version)
-     "cp2_products_idx2" btree (product_id)
- Foreign-key constraints:
-     "cp2_products_fk1" FOREIGN KEY (derived_product_uuid) REFERENCES cp2_products(uuid)
- Referenced by:
-     TABLE "cp2_owner_products" CONSTRAINT "cp2_owner_products_fk2" FOREIGN KEY (product_uuid) REFERENCES cp2_products(uuid)
-     TABLE "cp2_product_dependent_products" CONSTRAINT "cp2_proddepprod_fk1" FOREIGN KEY (product_uuid) REFERENCES cp2_products(uuid) ON DELETE CASCADE
-     TABLE "cp2_product_attributes" CONSTRAINT "cp2_product_attributes_fk1" FOREIGN KEY (product_uuid) REFERENCES cp2_products(uuid) ON DELETE CASCADE
-     TABLE "cp2_product_branding" CONSTRAINT "cp2_product_branding_fk1" FOREIGN KEY (product_uuid) REFERENCES cp2_products(uuid) ON DELETE CASCADE
-     TABLE "cp2_product_certificates" CONSTRAINT "cp2_product_certificates_fk1" FOREIGN KEY (product_uuid) REFERENCES cp2_products(uuid) ON DELETE CASCADE
-     TABLE "cp2_product_content" CONSTRAINT "cp2_product_content_fk1" FOREIGN KEY (product_uuid) REFERENCES cp2_products(uuid) ON DELETE CASCADE
-     TABLE "cp2_products" CONSTRAINT "cp2_products_fk1" FOREIGN KEY (derived_product_uuid) REFERENCES cp2_products(uuid)
-     TABLE "cp2_product_provided_products" CONSTRAINT "cp2_products_prov_prods_fk1" FOREIGN KEY (product_uuid) REFERENCES cp2_products(uuid) ON DELETE CASCADE
-     TABLE "cp2_product_provided_products" CONSTRAINT "cp2_products_prov_prods_fk2" FOREIGN KEY (provided_product_uuid) REFERENCES cp2_products(uuid) ON DELETE CASCADE
+                              Table "public.cp_products"
+         Column        |            Type             | Collation | Nullable | Default 
+ ----------------------+-----------------------------+-----------+----------+---------
+  uuid                 | character varying(32)       |           | not null | 
+  created              | timestamp without time zone |           |          | 
+  updated              | timestamp without time zone |           |          | 
+  namespace            | character varying(32)       |           | not null | 
+  product_id           | character varying(32)       |           | not null | 
+  name                 | character varying(255)      |           | not null | 
+  multiplier           | bigint                      |           |          | 
+  derived_product_uuid | character varying(32)       |           |          | 
+ Indexes:
+     "cp_products_pk" PRIMARY KEY, btree (uuid)
+     "cp_products_uidx1" UNIQUE CONSTRAINT, btree (namespace, product_id)
+     "cp_products_idx1" btree (namespace)
+     "cp_products_idx2" btree (product_id)
+ Foreign-key constraints:
+     "cp_products_fk1" FOREIGN KEY (derived_product_uuid) REFERENCES cp_products(uuid) ON DELETE SET NULL
+ Referenced by:
+     TABLE "cp_product_attributes" CONSTRAINT "cp_product_attributes_fk1" FOREIGN KEY (product_uuid) REFERENCES cp_products(uuid) ON DELETE CASCADE
+     TABLE "cp_product_branding" CONSTRAINT "cp_product_branding_fk1" FOREIGN KEY (product_uuid) REFERENCES cp_products(uuid) ON DELETE CASCADE
+     TABLE "cp_product_certificates" CONSTRAINT "cp_product_certificates_fk1" FOREIGN KEY (product_uuid) REFERENCES cp_products(uuid) ON DELETE CASCADE
+     TABLE "cp_product_content" CONSTRAINT "cp_product_content_fk1" FOREIGN KEY (product_uuid) REFERENCES cp_products(uuid) ON DELETE CASCADE
+     TABLE "cp_product_dependent_products" CONSTRAINT "cp_product_dps_fk1" FOREIGN KEY (product_uuid) REFERENCES cp_products(uuid) ON DELETE CASCADE
+     TABLE "cp_product_provided_products" CONSTRAINT "cp_product_provided_products_fk1" FOREIGN KEY (product_uuid) REFERENCES cp_products(uuid) ON DELETE CASCADE
+     TABLE "cp_product_provided_products" CONSTRAINT "cp_product_provided_products_fk2" FOREIGN KEY (provided_product_uuid) REFERENCES cp_products(uuid) ON DELETE CASCADE
+     TABLE "cp_products" CONSTRAINT "cp_products_fk1" FOREIGN KEY (derived_product_uuid) REFERENCES cp_products(uuid) ON DELETE SET NULL
```

File: src/main/java/org/candlepin/resteasy/filter/StoreFactory.java
Comment (Ceiu) [reply src/main/java/org/candlepin/resteasy/filter/StoreFactory.java]: Local testing suggests it was safe to remove entirely.

File: src/main/java/org/candlepin/controller/PoolManager.java
Comment (Ceiu) [reply src/main/java/org/candlepin/controller/PoolManager.java]: Oops, misread the class file; thought it was in PoolResource.

Added the query and some tests for this logic.

**Review by Januson — APPROVED** [review]

File: src/main/java/org/candlepin/resource/ProductResource.java
Comment (nikosmoum) [reply src/main/java/org/candlepin/resource/ProductResource.java:104]: I added a task in the backlog to get this paginated

File: src/main/java/org/candlepin/resource/OwnerProductResource.java
Comment (nikosmoum) [inline src/main/java/org/candlepin/resource/OwnerProductResource.java:571]: I didnt have a chance to fully look through the paging utility PR, but do I understand correctly that it applies paging after the fact/db query? So the db query may return say, 10000 results, all of those are fetched into memory, and then this util returns only the page requested by the client, so e.g. items 201-300?
If so, isn't this problematic memory-wise for endpoints/queries that fetch a lot of results, and paging at the query level was supposed to be a solution here?

Comment (nikosmoum) [reply src/main/java/org/candlepin/resource/OwnerProductResource.java:571]: Also, this is not really a problem for this specific endpoint, since based on past investigation I see will not return enough products to DoS candlepin, but for a few other endpoints it will be

File: src/main/java/org/candlepin/resource/OwnerContentResource.java
Comment (nikosmoum) [inline src/main/java/org/candlepin/resource/OwnerContentResource.java:294]: This is the kind of endpoint paging at the query level would help: right now there are 12k+ contents in existence, and we know some of those massive orgs have 8-9k of them. This endpoint would fetch all of them in memory, only to apply paging after the fact

**Review by nikosmoum — CHANGES_REQUESTED** [review]

File: src/main/java/org/candlepin/resource/OwnerProductResource.java
Comment (Ceiu) [reply src/main/java/org/candlepin/resource/OwnerProductResource.java:571]: Yes, your understanding of how the pagingutil works is correct; and it is not meant to be a replacement for properly mapping and narrowing queries in areas where we've determined the output needs to be limited (consumer fetching, for instance).

Given where the JPA interfaces are combined with our current timelines, I don't foresee us having the time to spec out some magic that can (a) translate and validate user-facing sortBy fields to internal model field names, (b) apply the query limitations generally in a way that we wouldn't need to hand-craft individual queries, and (c) get the total element count as a separate query, again, without needing to hand-craft it for the operation at hand.

The interface here is general enough that we can use it to apply paging quickly in places where the base query is safe enough that we allow it to be used without paging (of which, products and content should be), but not in heavier places where our fetch-all query is going to return 6+ figures of results.

File: src/main/java/org/candlepin/model/ProductContent.java
Comment (Ceiu) [reply src/main/java/org/candlepin/model/ProductContent.java:102]: Not without a larger refactor of this area and some thought/investigation/design around how we want that to work. It's an echo of permitting direct-assigning collections of product-content join objects directly on the target product.

File: src/main/java/org/candlepin/model/ProductCurator.java
Comment (Ceiu) [reply src/main/java/org/candlepin/model/ProductCurator.java:447]: Fixed

File: src/main/java/org/candlepin/model/ProductCurator.java
Comment (Ceiu) [reply src/main/java/org/candlepin/model/ProductCurator.java:468]: Fixed

File: src/main/java/org/candlepin/resource/OwnerContentResource.java
Comment (Ceiu) [reply src/main/java/org/candlepin/resource/OwnerContentResource.java:294]: True enough, and a good warning; but with this refactor, those content numbers will get paired down to ~12k total.

File: src/main/java/org/candlepin/controller/refresher/RefreshWorker.java
Comment (nikosmoum) [inline src/main/java/org/candlepin/controller/refresher/RefreshWorker.java:80]: With this change, what happens when a manifest consumer's entitlement expires, and subsequently the corresponding pool is deleted during the manifest import, and no other pool is using the product/content that this pool provided? The products/content stay in the database because they are in the global namespace?

Comment (Ceiu) [reply src/main/java/org/candlepin/controller/refresher/RefreshWorker.java:80]: Correct. Refresh/import will no longer delete products or contents for any reason (again).

I was skeptical of that at first thought, but given that (a) hosted disables orphan cleanup entirely and (b) sat wants a long lead time on it as well (30 days from detection + next import IIRC), it seems like something users don't actually want anyway. Given the substantial reduction in data usage, I don't expect anyone will actually care; and if they do, we can revisit this in a branch specifically for Sat.

**Review by nikosmoum — APPROVED** [review]

**Pull request merged**
